File: binary_search/367_valid_perfect_square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num < 2: return True
        
        left, right = 2, num // 2
        
        while left <= right:
            mid = (left + right) // 2
            
            res = mid * mid
            if res == num:
                return True
            elif res > num:
                right = mid - 1
            else:
                left = mid + 1
        return False
        
        

File: binary_search/367_valid_perfect_square.java
class Solution {
    public boolean isPerfectSquare(int num) {
        long left = 2, right = num / 2;

        while (left <= right) {
            long mid = left + (right - left) / 2;
            long square = mid * mid;

            if (square == num) {
                return true;
            } else if (square > num) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return true ? num < 2 : false;
    }
}